pkg/compact/dedup/replica.go

```diff
@@ -0,0 +1,258 @@
+		return names[i] < names[j]
+	})
+	var b strings.Builder
+	b.WriteString("[")
+	for i, name := range names {
+		if i != 0 {
+			b.WriteString(",")
+		}
+		b.WriteString(fmt.Sprintf("%s=%s", name, labels[name]))
+	}
+	b.WriteString("]")
+	return b.String()
+}
+
+type ReplicaSyncer struct {
+	logger               log.Logger
+	metrics              *DedupMetrics
+	bkt                  objstore.Bucket
+	consistencyDelay     time.Duration
+	blockSyncConcurrency int
+
+	labelName string
+	mtx       sync.Mutex
+	blocksMtx sync.Mutex
+}
+
+func NewReplicaSyncer(logger log.Logger, metrics *DedupMetrics, bkt objstore.Bucket, labelName string,
+	consistencyDelay time.Duration, blockSyncConcurrency int) *ReplicaSyncer {
+	return &ReplicaSyncer{
+		logger:               logger,
+		metrics:              metrics,
+		bkt:                  bkt,
+		labelName:            labelName,
+		consistencyDelay:     consistencyDelay,
+		blockSyncConcurrency: blockSyncConcurrency,
+	}
+}
+
+func (s *ReplicaSyncer) Sync(ctx context.Context) (Replicas, error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	blocks := make(map[ulid.ULID]*metadata.Meta)
+	metaIdsChan := make(chan ulid.ULID)
+	errChan := make(chan error, s.blockSyncConcurrency)
+
+	syncCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for i := 0; i < s.blockSyncConcurrency; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for id := range metaIdsChan {
+				s.blocksMtx.Lock()
+				_, seen := blocks[id]
+				s.blocksMtx.Unlock()
+				if seen {
+					continue
+				}
+				s.metrics.syncMetas.WithLabelValues(s.bkt.Name()).Inc()
+				begin := time.Now()
+				meta, err := s.download(syncCtx, id)
+				s.metrics.syncMetaDuration.WithLabelValues(s.bkt.Name()).Observe(time.Since(begin).Seconds())
+				if err != nil {
+					errChan <- err
+					s.metrics.syncMetaFailures.WithLabelValues(s.bkt.Name(), id.String()).Inc()
+					return
+				}
+				if meta == nil {
+					continue
+				}
+				s.blocksMtx.Lock()
+				blocks[id] = meta
+				s.blocksMtx.Unlock()
+			}
+		}()
+	}
+
+	remote := map[ulid.ULID]struct{}{}
+
+	err := s.bkt.Iter(ctx, "", func(name string) error {
+		id, ok := block.IsBlockDir(name)
+		if !ok {
+			return nil
+		}
+
+		remote[id] = struct{}{}
+
+		select {
+		case <-ctx.Done():
+		case metaIdsChan <- id:
+		}
+		return nil
+	})
+
+	close(metaIdsChan)
+
+	if err != nil {
+		return nil, compact.Retry(errors.Wrapf(err, "sync block metas from bucket %s", s.bkt.Name()))
+	}
+
+	wg.Wait()
+	close(errChan)
+
+	if err := <-errChan; err != nil {
+		return nil, compact.Retry(errors.Wrapf(err, "download block metas from bucket %s", s.bkt.Name()))
+	}
+
+	result := make([]*metadata.Meta, 0)
+	for id, b := range blocks {
+		if _, ok := remote[id]; ok {
+			result = append(result, b)
+		}
+	}
+	return NewReplicas(s.labelName, result)
+}
+
+func (s *ReplicaSyncer) download(ctx context.Context, id ulid.ULID) (*metadata.Meta, error) {
```

Comment:
Not a fan of adding yet another Download method. Same with Syncer. Can we reuse those?

We already have problem with having multiple Syncer so we have https://github.com/thanos-io/thanos/issues/1335

Reply:
Sure, will figure out one proper way to reuse the functions in the `compact.Syncer`.
From the #1335, I saw we would start to unify the syncer APIs.  
I think if I can do this change after #1335, it will save some efforts.